## Concepts Covered
- [Standard Error of the Sample Mean](#standard-error-of-the-sample-mean)
- [Z-Score and Standartization](#Z-Score-and-Standartization)
- [Confidence Level and Confidence Interval](#confidence-level-and-confidence-interval)

In [22]:
import pandas as pd
import seaborn as sns
import scipy.stats as st
import math
import statsmodels.stats.api as sm
from sklearn import preprocessing
import numpy as np



In [2]:
data_path = "Inf_Stats_files/Confidence Intervals/pizza_restaurant.csv"
df = pd.read_csv(data_path)
df.head()

,Product Name,Crust,Toppings,Price,Delivery Time,# pizzas the customer ordered before
0,Pepperoni,Cheese_and_Garlic,4,17,26.3,4
1,Hawaiian,Cheese_and_Chili,4,17,27.8,4
2,Calzone,Cheese_and_Garlic,3,20,31.5,7
3,Margherita,Cheese,4,23,20.8,7
4,Calzone,Cheese_and_Garlic,4,19,27.7,8


In [3]:
df.describe()

,Toppings,Price,Delivery Time,# pizzas the customer ordered before
count,1000.000000,1000.000000,1000.000000,1000.000000
mean,3.965000,19.342000,25.061100,5.659000
std,1.021185,3.345479,2.490397,2.459831
min,1.000000,12.000000,17.800000,0.000000
25%,3.000000,17.000000,23.300000,4.000000
50%,4.000000,19.000000,25.100000,5.000000
75%,5.000000,21.000000,26.700000,7.000000
max,7.000000,33.000000,32.400000,15.000000


# Standard Error of the Sample Mean
It estimates how far the sample mean is likely to be from the population mean.

Standard Error is calculated as follows:
$$
SE = \frac{\sigma}{\sqrt{n}}
$$
where $\sigma$ is standard deviation of the sample, $n$ is number of observations.

[Back to top](#concepts-covered)


In [4]:
# SE of price
# through formula:
print(df.Price.std() / math.sqrt(df.Price.count())) 
# with stats package:
print(st.sem(df.Price))

0.105793327900337
0.105793327900337


# Z-Score and Standartization

Z-score estimates how far away a data point is from its mean.
$$
z = \frac{x - \mu}{\sigma}
$$

[Back to top](#concepts-covered)


In [5]:
# Z-score for delivery time
# using formula
df["Delivery Time standardized"] = (df["Delivery Time"] - df["Delivery Time"].mean())/df["Delivery Time"].std()

In [6]:
# using sklearn
df["Delivery Time standardized2"] = preprocessing.scale(df["Delivery Time"])

In [7]:
# Comapre the results
df.head()

,Product Name,Crust,Toppings,Price,Delivery Time,# pizzas the customer ordered before,Delivery Time standardized,Delivery Time standardized2
0,Pepperoni,Cheese_and_Garlic,4,17,26.3,4,0.497471,0.497720
1,Hawaiian,Cheese_and_Chili,4,17,27.8,4,1.099784,1.100335
2,Calzone,Cheese_and_Garlic,3,20,31.5,7,2.585491,2.586785
3,Margherita,Cheese,4,23,20.8,7,-1.711012,-1.711868
4,Calzone,Cheese_and_Garlic,4,19,27.7,8,1.059630,1.060160


In [8]:
df.describe()

,Toppings,Price,Delivery Time,# pizzas the customer ordered before,Delivery Time standardized,Delivery Time standardized2
count,1000.000000,1000.000000,1000.000000,1000.000000,1.000000e+03,1.000000e+03
mean,3.965000,19.342000,25.061100,5.659000,1.385558e-16,1.421085e-16
std,1.021185,3.345479,2.490397,2.459831,1.000000e+00,1.000500e+00
min,1.000000,12.000000,17.800000,0.000000,-2.915639e+00,-2.917098e+00
25%,3.000000,17.000000,23.300000,4.000000,-7.071562e-01,-7.075101e-01
50%,4.000000,19.000000,25.100000,5.000000,1.562000e-02,1.562781e-02
75%,5.000000,21.000000,26.700000,7.000000,6.580878e-01,6.584171e-01
max,7.000000,33.000000,32.400000,15.000000,2.946879e+00,2.948354e+00


In [9]:
# Values vary due to the difference in standard deviation calculated.
# To obtain the same result as in sklearn.preprocessing.scale(), in the formula # we need to pass ddof=0 to use standard deviation of the population.

df["Delivery Time standardized"] = (df["Delivery Time"] - df["Delivery Time"].mean())/df["Delivery Time"].std(ddof=0)
df.describe()

,Toppings,Price,Delivery Time,# pizzas the customer ordered before,Delivery Time standardized,Delivery Time standardized2
count,1000.000000,1000.000000,1000.000000,1000.000000,1.000000e+03,1.000000e+03
mean,3.965000,19.342000,25.061100,5.659000,1.421085e-16,1.421085e-16
std,1.021185,3.345479,2.490397,2.459831,1.000500e+00,1.000500e+00
min,1.000000,12.000000,17.800000,0.000000,-2.917098e+00,-2.917098e+00
25%,3.000000,17.000000,23.300000,4.000000,-7.075101e-01,-7.075101e-01
50%,4.000000,19.000000,25.100000,5.000000,1.562781e-02,1.562781e-02
75%,5.000000,21.000000,26.700000,7.000000,6.584171e-01,6.584171e-01
max,7.000000,33.000000,32.400000,15.000000,2.948354e+00,2.948354e+00


# Confidence Level and Confidence Interval

<b>Confidence level</b> indicates how sure one can be that a statistical conclusion is correct. Expressed in $\%$.

A typical confidence level is set at 95$\%$ chance that the results are accurate.

Confidence level is used in hypothesis testing and is used to calculate confidence intervals.

High confidence level doesn't mean higher accuracy, just higher certainty.

<b>Confidence interval</b> is a range likely to contain the true population parameter with a certain level certainty.

Confidence interval width is determined by:
- Standard error of the sample mean
- Confidence level

[Back to top](#concepts-covered)

In [10]:
cl = [0.005, 0.025, 0.05, 0.95, 0.975, 0.995]
for alpha in cl:
    print(f"The corresponding value in normal distribution for {alpha} is: {round(st.norm.ppf(alpha), 3)}")

The corresponding value in normal distribution for 0.005 is: -2.576
The corresponding value in normal distribution for 0.025 is: -1.96
The corresponding value in normal distribution for 0.05 is: -1.645
The corresponding value in normal distribution for 0.95 is: 1.645
The corresponding value in normal distribution for 0.975 is: 1.96
The corresponding value in normal distribution for 0.995 is: 2.576


In [11]:
# Inverse of ppf:
st.norm.cdf(1.96)

np.float64(0.9750021048517795)

<b>Confidence Intervals for Large Samples</b>

$$
CI = \bar{x} \pm z \frac{s}{\sqrt{n}}
$$

Sample size must be above 30 (ideally above 100) observations.

Practical advice: not to make conclusions on samples of 10 or less observations.

Z-value represents the desired confidence level (often 95%). Confidence interval: from 2.5 to 97.5 percentile. To determine CI, find z value corresponding to 2.5 or 97.5% instead of 95%

<i>Example.</i>

A survey shows that 60% of customers would buy the new product. The company wants to know the survey's accuracy. Confidence interval can estimate the range of possible true populaton percentages. For instance, 95% confidence interval of 55-65% would mean 95% confidence that the true percentage of customers falls within that range. This helps recogizing the level of uncertainty in the predictions and evaluate the potential risks.

In [14]:
# Confidence interval for the Price mean
print(f"The mean is {df.Price.mean()}")
print(f"The 95-confidence interval for it: {st.norm.interval(confidence=0.95,
                                            loc = df.Price.mean(),
                                            scale = st.sem(df.Price))}")

The mean is 19.342
The 95-confidence interval for it: (np.float64(19.134648887510703), np.float64(19.549351112489294))


In [19]:
# function to compute the confidence interval for the mean
# for each numerical variable that has a sample size bigger than 30
def conf_interval_for_mean(X: pd.Series, confidence=0.95)->tuple[float, float]:
    if not isinstance(X, pd.Series):
        raise TypeError("X must be a pandas Series.")
    if not 0 < confidence < 1:
        raise ValueError("confidence must be between 0 and 1.")
    if X.count() < 30:
        raise ValueError(f"At least 30 non-null observations are required, got {X.count()}.")

    x = X.dropna()
    mean = x.mean()
    se = st.sem(x)

    ci_low, ci_high = st.norm.interval(confidence=confidence, loc=mean, scale=se)
    return (float(ci_low), float(ci_high))

In [25]:
def calculate_confidence_intervals(df, confidence=0.95):
    intervals = {}

    # Iterate over each numerical column in the dataframe
    for column in df.select_dtypes(include=[np.number]).columns:
        data = df[column].dropna()
        # Check if the sample size is greater than 30
        if len(data) > 30:
            # Compute the confidence interval for the mean
            interval = st.norm.interval(confidence, loc=data.mean(), scale=st.sem(data))
            intervals[column] = interval
    return intervals


print(calculate_confidence_intervals(df))

{'Toppings': (np.float64(3.9017074909279676), np.float64(4.028292509072032)), 'Price': (np.float64(19.134648887510703), np.float64(19.549351112489294)), 'Delivery Time': (np.float64(24.9067464105456), np.float64(25.2154535894544)), '# pizzas the customer ordered before': (np.float64(5.5065408812039385), np.float64(5.811459118796061)), 'Delivery Time standardized': (np.float64(-0.06201051624377051), np.float64(0.06201051624377079)), 'Delivery Time standardized2': (np.float64(-0.06201051624377051), np.float64(0.06201051624377079))}
